starlasu-client/src/main/kotlin/KolasuClient.kt

```diff
@@ -5,25 +5,103 @@ import com.strumenta.kolasu.lionweb.ConstantSourceIdProvider
 import com.strumenta.kolasu.lionweb.KNode
 import com.strumenta.kolasu.lionweb.LionWebModelConverter
 import com.strumenta.kolasu.lionweb.LionWebNodeIdProvider
+import com.strumenta.kolasu.lionweb.LionWebPartition
 import com.strumenta.kolasu.lionweb.PrimitiveValueSerialization
+import com.strumenta.kolasu.lionweb.SimpleSourceIdProvider
 import com.strumenta.kolasu.lionweb.SourceIdProvider
 import com.strumenta.kolasu.lionweb.StructuralLionWebNodeIdProvider
 import com.strumenta.kolasu.model.Node
+import com.strumenta.kolasu.model.SyntheticSource
 import com.strumenta.kolasu.model.children
 import com.strumenta.kolasu.model.containingProperty
 import com.strumenta.kolasu.model.indexInContainingProperty
+import com.strumenta.kolasu.traversing.walk
 import com.strumenta.lwrepoclient.base.LionWebClient
 import io.lionweb.lioncore.java.serialization.JsonSerialization
 import io.lionweb.lioncore.java.utils.CommonChecks
 import java.io.File
+import java.util.IdentityHashMap
 import kotlin.reflect.KClass
 import kotlin.reflect.KProperty1
 
+class DefaultLionWebRepositoryNodeIdProvider(var sourceIdProvider: SourceIdProvider = SimpleSourceIdProvider()) :
+    LionWebNodeIdProvider {
+    protected open fun partitionId(kNode: Node): String {
+        require(kNode.parent == null)
+        require(kNode.source != null)
+        // TODO update SimpleSourceIdProvider
+        if (kNode.source is SyntheticSource) {
+            return "synthetic_" + (kNode.source as SyntheticSource).description
+        }
+        return sourceIdProvider.sourceId(kNode.source)
+    }
+
+    override fun id(kNode: Node): String {
+        if (kNode.isPartition) {
+            return partitionId(kNode)
+        }
+        val id = "${sourceIdProvider.sourceId(kNode.source)}_${kNode.positionalID}"
+        if (!CommonChecks.isValidID(id)) {
+            throw IllegalStateException("An invalid LionWeb Node ID has been produced")
+        }
+        return id
+    }
+
+    private val KNode.positionalID: String
+        get() {
+            return if (this.parent == null) {
+                "root"
+            } else {
+                val cp = this.containingProperty()!!
+                val postfix = if (cp.multiple) "${cp.name}_${this.indexInContainingProperty()!!}" else cp.name
+                "${this.parent!!.positionalID}_$postfix"
+            }
+        }
+}
+
+/**
+ * This NodeIdProvider remembers IDs explicitly set. This is useful because we often insert nodes in an existing
+ * tree. In the moment we do the insertion we know where the Node will end up and we can calculate the ID based on
+ * that. Later, if we access that Node without that contextual information we would not be able to recalculate the
+ * same Node ID.
+ */
+class OverridableNodeIdProvider(private val kolasuClient: KolasuClient) : LionWebNodeIdProvider {
```

Comment:
Does it make sense to have a way for cleaning the overrides, or a strategy etc.? What's the reasoning behind this, is one supposed to replace the id provider when they no longer want to hold onto old nodes?

Reply:
The reasons are two:
* One may want to specify explicitly IDs for certain partitions. Suppose you want to create a special partition to list known languages or to contain the original code and call them `languages` or `original-code`, instead of a Node produced by the NodeIdProvider logic. Without this, you would need to somehow hack your NodeIdProvider
* When we insert an AST in a partition, we will obtaining IDs based on the ID of the partition. This is known at insertion time and therefore we can calculate the proper LionWeb Node IDs. However, if you later want to get the Node ID associated to any node of the AST, you could recalculate it only passing the partition where you inserted, the containment and the index at which you inserted them. This would be cumbersome. In this way instead the KolasuClient will remember the Node IDs, so that it is easy for you to later retrieving them

I am not sure if we could add a strategy to clear the caches automatically. Perhaps we can offer an hook to do it explicitly.

Reply:
So the partition is not an AST but a container of ASTs?

Reply:
In the development version of the code insight studio, which uses an older model repository client, a partition is the same thing that you persist using `storeTree`, that's why I assumed it to be an AST root.

Reply:
With the old APIs, I would have also used ASTs as partitions, however we need to change this because of the new APIs.

I wrote about some reasons here: https://www.notion.so/strumenta/Using-LionWeb-Repository-for-storing-StarLasu-ASTs-ff278d06da304304991cce61d58a065d?pvs=4#60df5c4ba8964efcbf1880cd2871c037

I think that it would make sense to create specific concepts that we use to represent partitions such `Codebase`, which can have as children the `SourceFile` concepts that you planned to introduce in a language specific for the Code Insight Studio.